tao_compiler/mlir/disc/transforms/disc_transform_schedule.cc

```diff
@@ -0,0 +1,430 @@
+
+  // %fill = transform.structured.match ops{["linalg.fill"]} in %variant
+  Value fill = buildMatchOp(b, loc, variant, {"linalg.fill"});
+  // %matmul = transform.structured.match ops{["linalg.matmul"]} in %arg1
+  Value matmul = buildMatchOp(b, loc, variant, {"linalg.matmul"});
+
+  // transform.structured.tile_to_foreach_thread_op %matmul num_threads [1, 1]
+  auto forEachThreadOp = buildTileToForEachThreadOp(b, loc, matmul, {1, 1});
+  Value forEachThreadLoop = forEachThreadOp->getResult(0);
+  Value tiledMatmul = forEachThreadOp->getResult(1);
+
+  // transform.structured.fuse_into_containing_op %fill into %0#0
+  auto fuseIntoContainingOp =
+      buildFuseIntoContainingOp(b, loc, fill, forEachThreadLoop);
+
+  // first level tile and fuse matmul and fill op.
+  auto fuseOp0 = buildFuseOp(b, loc, tiledMatmul, {288, 48, 0}, {0, 1, 2});
+
+  // second level tile and fuse matmul and fill op.
+  auto fuseOp1 =
+      buildFuseOp(b, loc, fuseOp0->getResult(0), {6, 16, 0}, {0, 1, 2});
+
+  // gemm reduction axis tiling
+  auto tileOp =
+      buildTileOp(b, loc, fuseOp1->getResult(0), {0, 0, 1}, {0, 1, 2});
+
+  variant = buildRunCanonicalizer(b, loc, variant);
+
+  // fold two extract_slice ops generated by two-level tiling. It's needed to
+  // enable following pad and hosit schedule.
+  Value weightInnerSlice =
+      buildGetProducerOfOperand(b, loc, tileOp->getResult(0), 1);
+  buildFoldProducerExtractSlice(b, loc, weightInnerSlice, 2);
+
+  // pad to match the requirement of hardware vector/tensor instruction.
+  auto padOp = buildPadOp(b, loc, tileOp->getResult(0), {0, 1, 2});
+
+  Value padForInput = buildGetProducerOfOperand(b, loc, padOp, 0);
+  Value padForWeight = buildGetProducerOfOperand(b, loc, padOp, 1);
+  forEachThreadLoop = buildMatchOp(b, loc, variant, {"scf.foreach_thread"});
+  auto outterLoopForN = buildGetParentForOp(b, loc, padForInput, 4);
+  buildCacheRead(b, loc, padForWeight, forEachThreadLoop, {1, 1}, {1, 16}, true,
+                 {2, 0, 1, 3});
+  buildCacheRead(b, loc, padForInput, outterLoopForN, {1, 1}, {6, 1}, true,
+                 {0, 2, 3, 1});
+
+  variant = buildRunCanonicalizer(b, loc, variant);
+
+  Value multiLevelPackOps =
+      buildMatchOp(b, loc, variant, {"disc_linalg_ext.multi_level_pack"});
+  buildLowerMultiLevelPackToLoop(b, loc, multiLevelPackOps);
+
+  variant = buildRunCanonicalizer(b, loc, variant);
+
+  Value func = buildMatchOp(b, loc, variant, {"func.func"});
+  buildVectorize(b, loc, func, true);
+
+  variant = buildRunCanonicalizer(b, loc, variant);
+  variant = buildDISCBufferize(b, loc, variant);
+
+  buildLowerVectors(b, loc, {0, 1, 2, 3, 4}, "outerproduct", "innerparallel",
+                    "linalg-copy", true, "eltwise", false);
+  buildLowerVectors(b, loc, {5, 6, 7}, "outerproduct", "innerparallel",
+                    "linalg-copy", true, "eltwise", false);
+  b.create<transform::YieldOp>(loc);
+  return success();
+}
+
+DISC_TRANSFORM_SCHEDULE(PatternKind::kGEMM, "",
```

Comment:
Thanks, I'll add it.

Reply:
I think more about this. Maybe we should just use one tag instead of adding an explicit filed for device.

For example we may have different schedule for x86 and aarch64. And even within the same cpu family, there are many different hardware features. thus using tag maybe more robust.

Reply:
As discussed offline, let leave a single `tag` here, and refine it if necessary when doing scheduling selection.